src/main/java/cn/nukkit/block/BlockFire.java

```diff
@@ -86,8 +111,12 @@ public void onEntityCollide(Entity entity) {
         return new Item[0];
     }
 
+    @PowerNukkitDifference(info = "Soul Fire Implementation", since = "1.4.0.0-PN")
     @Override
     public int onUpdate(int type) {
+        if ((this.down().getId() == Block.SOUL_SAND ||  this.down().getId() == Block.SOUL_SOIL) && this.getId() == BlockID.FIRE){
+            this.getLevel().setBlock(this, Block.get(BlockID.SOUL_FIRE));
+        }
```

Comment:
Needs to check if `type == Level.BLOCK_UPDATE_NORMAL`, skipping this check may cause a lot of random TPS lag.

`this.down()` may be cached and reused in a common variable, avoiding having to acquire synchronization locks multiple times to get the same data.